### PR TITLE
rust: cargo: Use rustc and cargo built on Build for cross-compiling

### LINKED ISSUE #207435: pkgsStatic.xxx with Rust requires to recompile rustc
### Describe the bug
When attempting to statically build a rust package, it needs to compile `rustc` first and doesn't get it from the cache.

### Steps To Reproduce
`nix build nixpkgs#pkgsStatic.bandwhich`

### Expected behavior
It would be expected to fetch the `rustc` from the cache, and use that to build the rust-based package.

### PR DESCRIPTION
**Pull request by mehmooda**
When cross-compiling a rust package, all we need is the std library compiled
for the target. This uses the final stage compiler which was built for Build
and then uses that as a stage0 compiler for target std library.

It also copies the rust binary from pkgsBuildBuild so that it find the new
lib/rustlib directory.

We also need to create a cargo wrapper which will use the "new" rust compiler

Closes #207435.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/rust/cargo_cross.nix
Comment (alyssais) [inline pkgs/development/compilers/rust/cargo_cross.nix:4]: `runCommand` is probably more suitable for this.  Should also inherit most of `meta`, and link to all the non-executable stuff like man pages, because people will want to use the cross cargo in nix shells.

File: pkgs/development/compilers/rust/default.nix
Comment (Mindavi) [inline pkgs/development/compilers/rust/default.nix:22]: I think we use camelCase for options

File: pkgs/development/compilers/rust/cargo_cross.nix
Comment (alyssais) [inline pkgs/development/compilers/rust/cargo_cross.nix:21]: Maybe it would be better to just re-wrap `${pkgsBuildBuild.cargo}/bin/cargo`?  That way you wouldn't need to specify the SSL, which I'm worried would otherwise go out of sync.

File: pkgs/development/compilers/rust/cargo_cross.nix
Comment (alyssais) [inline pkgs/development/compilers/rust/cargo_cross.nix:2]:
```suggestion
{ stdenv, pkgsBuildBuild, makeShellWrapper, cacert, rustc }:
```

File: pkgs/development/compilers/rust/cargo_cross.nix
Comment (alyssais) [inline pkgs/development/compilers/rust/cargo_cross.nix:1]:
```suggestion
```

File: pkgs/development/compilers/rust/default.nix
Comment (alyssais) [inline pkgs/development/compilers/rust/default.nix:21]: I'm not sure this should be overrideable, since it presumably wouldn't work in any other case, and shouldn't ever be the wrong thing to do in this case.

File: pkgs/development/compilers/rust/rustc.nix
Comment (alyssais) [inline pkgs/development/compilers/rust/rustc.nix:13]:
```suggestion
, fastCross
```

Otherwise it's inconsistent with every other option / variable.

File: pkgs/development/compilers/rust/rustc.nix
Comment (alyssais) [inline pkgs/development/compilers/rust/rustc.nix:164]: As with cargo, we should make sure documentation etc is still accessible in the cross-compiler versions.

**Review by alyssais — DISMISSED** [review]

File: pkgs/development/compilers/rust/rustc.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/rust/rustc.nix:114]:
```suggestion

  # if we already have a rust compiler for build just compile the target std library and reuse compiler
  buildPhase = lib.optionalString fastCross "
    python x.py --stage=0 build library/std
  ";
```

File: pkgs/development/compilers/rust/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/rust/default.nix:26]:
```suggestion
  fastCross = (stdenv.buildPlatform == stdenv.hostPlatform) && (stdenv.hostPlatform != stdenv.targetPlatform);
```

a short comment here wouldn't hurt.

File: pkgs/development/compilers/rust/cargo_cross.nix
Comment (mehmooda) [reply pkgs/development/compilers/rust/cargo_cross.nix:13]: a single symbolic link of the share folder seems to work. I'm not sure that a recursive shadow copy is necessary.

File: pkgs/development/compilers/rust/rustc.nix
Comment (mehmooda) [reply pkgs/development/compilers/rust/rustc.nix]: Done.

File: pkgs/development/compilers/rust/cargo_cross.nix
Comment (mehmooda) [reply pkgs/development/compilers/rust/cargo_cross.nix]: I've updated to runCommand, and inherited meta

File: pkgs/development/compilers/rust/rustc.nix
Comment (Mindavi) [inline pkgs/development/compilers/rust/rustc.nix:112]: Make sure to call the hooks here and in the installPhase.

**Comment by Ericson2314** [comment]
I think this sort of thing is good to have.

**Comment by mehmooda** [comment]
I think this is ready now.
I'll queue up a clean up to staging for the if else null into a lib.optionalString after this and master gets merged into staging.

**Comment by mehmooda** [comment]
TIL: if you ln -s a directory straight onto the output, it works when nix is installed single user but fails when nix is running as a daemon due to it trying to chown the existing directory. Using lndir instead works.

**Comment by 06kellyjac** [comment]
anything else holding this up other than conflicts?

**Comment by superherointj** [comment]
@mehmooda Please fix conflicts.

**Comment by yu-re-ka** [comment]
Rebased

**Comment by yu-re-ka** [comment]
Moved lndir to nativeBuildInputs

**Comment by yu-re-ka** [comment]
This does not work with the current version of Rust. The compiler even after being copied to the new location is still looking for core crate exclusively in its original path, and this seems to be the compiled in sysroot (`rustc --print sysroot`).
Once more a wakeup call for someone to come up with a rustc wrapper ;)

We can still work around this however, by rewriting this reference in the binary to the new $out. But for that, the derivation name length has to match.

**Comment by winterqt** [comment]
> Once more a wakeup call for someone to come up with a rustc wrapper ;)

If it's compiled in, how would a wrapper help?

**Comment by yu-re-ka** [comment]
It can be altered with the `--sysroot` parameter

**Comment by yu-re-ka** [comment]
cross-compiling `fd` fails with `error: failed to get bitcode from object file for LTO (Bitcode section not found in object file)`, presumably because of other assumptions that are compiled into the rustc binary

**Comment by yu-re-ka** [comment]
The std lib produced with fastCross is a stage 0 std lib, and thus does not include LLVM bitcode.

https://github.com/nnethercote/rust/blob/a105c5c2c01ea66f59bc6f3f9ade65016af56160/src/bootstrap/compile.rs#L235

**Comment by yu-re-ka** [comment]
This should now work. Please review.

**Review by kloenk — APPROVED** [review]
LGTM

**Comment by yu-re-ka** [comment]
@ofborg build pkgsCross.aarch64-multiplatform.fd

**Comment by yu-re-ka** [comment]
@ofborg build pkgsCross.aarch64-multiplatform.fd

**Comment by yu-re-ka** [comment]
@ofborg build pkgsCross.musl64.fd

**Comment by yu-re-ka** [comment]
@ofborg build pkgsStatic.fd

**Comment by yu-re-ka** [comment]
Note: pkgsStatic.fd on aarch64-linux breakage is unrelated; was already broken on master

**Comment by infinisil** [comment]
This doesn't appear to have fixed the linked issue, see https://github.com/NixOS/nixpkgs/issues/207435#issuecomment-1720304196. Ping @winterqt

**Comment by yu-re-ka** [comment]
That a rustc derivation appears in the list is to be expected. Within that derivation is should only build the std lib though.

15 Sept 2023 02:03:38 Silvan Mosberger ***@***.***>:

**Pull request merged**
